result.nim

```diff
@@ -128,53 +144,58 @@ type
 
     case isOk*: bool
     of false:
-      error*: E
+      error: E
     of true:
-      value*: T
+      value: T
 
-  # TODO cannot raise generic error, so cannot include error value :(
-  # https://github.com/nim-lang/Nim/issues/7845
-  ResultError = object of Exception
-
-proc ok*(r: typedesc[Result],  v: auto): r  {.inline.} =
+proc ok*(R: type Result, v: auto): R {.inline.} =
   ## Initialize a result with a success and value
   ## Example: `Result[int, string].ok(42)`
-  r(isOk: true, value: v)
+  R(isOk: true, value: v)
 
 proc ok*(self: var Result, v: auto) {.inline.} =
   ## Set the result to success and update value
   ## Example: `result.ok(42)`
   self = Result.ok(v)
 
-proc err*(r: typedesc[Result], e: auto): r  {.inline.} =
+proc err*(R: type Result, e: auto): R {.inline.} =
   ## Initialize the result to an error
   ## Example: `Result[int, string].err("uh-oh")`
-  r(isOk: false, error: e)
+  R(isOk: false, error: e)
 
-proc err*(self: var Result, v: auto)  {.inline.} =
+proc err*(self: var Result, v: auto) {.inline.} =
   ## Set the result as an error
   ## Example: `result.err("uh-oh")`
   self = Result.err(v)
 
 template isErr*(self: Result): bool = not self.isOk
 
-proc map*[T, E, A](self: Result[T, E], f: proc(x: T): A {.closure.}): Result[A, E]  {.inline.} =
+proc map*[T, E, A](
+    self: Result[T, E], f: proc(x: T): A): Result[A, E] {.inline.} =
```

Comment:
Nim is quite finicky when semchecking higher order functions.

It tends to require either extraneous nimcall/closure or gcsafe, locks information.

This might cause widespread breakage due to that.

Reply:
well, I can imagine going `untyped` or `auto` here - slightly less nice but perhaps less picky about typing?